src/analyzing/Constant_propagation.ml

```diff
@@ -345,72 +345,70 @@ let var_stats prog : var_stats =
-                                  id_resolved = { contents = Some (_kind, sid) };
-                                  _;
-                                } ));
-                        _;
-                      };
-                  ],
-                  _ ) ) ->
-              let var = (H.str_of_ident id, sid) in
-              let stat = get_stat_or_create var h in
-              incr stat.lvalue
-          | N (Id (id, { id_resolved = { contents = Some (_kind, sid) }; _ }))
-            ->
-              let var = (H.str_of_ident id, sid) in
-              let stat = get_stat_or_create var h in
-              incr stat.rvalue;
-              k x
-          | _ -> k x);
-    }
+  let visitor =
+    object (self : 'self)
+      inherit [_] AST_generic.iter_no_id_info as super
+
+      method! visit_definition env x =
+        match x with
+        | ( {
+              name =
+                EN
+                  (Id
+                    (id, { id_resolved = { contents = Some (_kind, sid) }; _ }));
+              _;
+            },
+            VarDef { vinit = Some _e; _ } ) ->
+            let var = (H.str_of_ident id, sid) in
+            let stat = get_stat_or_create var h in
+            incr stat.lvalue;
+            super#visit_definition env x
+        | _ -> super#visit_definition env x
+
+      method! visit_expr env x =
+        match x.e with
+        (* TODO: What if there is an asignment inside the `lhs` ? *)
+        | Assign (* v = ... *) (lhs, _, e2)
+        | AssignOp (* v += ... *) (lhs, _, e2) ->
+            lvars_in_lhs lhs
+            |> List.iter (fun (id, sid) ->
+                   let var = (H.str_of_ident id, sid) in
+                   let stat = get_stat_or_create var h in
+                   incr stat.lvalue;
+                   match x.e with
+                   | AssignOp _ -> incr stat.rvalue
+                   | _ -> ());
+            self#visit_expr env e2
+        | Call
+            ( { e = IdSpecial (IncrDecr _, _); _ },
+              ( _,
+                [
+                  Arg
+                    {
+                      e =
+                        N
+                          (Id
+                            ( id,
+                              {
+                                id_resolved = { contents = Some (_kind, sid) };
+                                _;
+                              } ));
+                      _;
+                    };
+                ],
+                _ ) ) ->
+            let var = (H.str_of_ident id, sid) in
+            let stat = get_stat_or_create var h in
+            incr stat.lvalue
+        | N (Id (id, { id_resolved = { contents = Some (_kind, sid) }; _ })) ->
+            let var = (H.str_of_ident id, sid) in
+            let stat = get_stat_or_create var h in
+            incr stat.rvalue;
+            super#visit_expr env x
+        | _ -> super#visit_expr env x
+    end
   in
-  let visitor = V.mk_visitor ~vardef_assign:false hooks in
```

Comment:
we don't need to set something equivalent to vardef_assign:false with this new visitor?
By default vardef_assign is set to false?

Reply:
Correct, it's set to `false` by default.

Reply:
But also, this is using a vanilla visitor rather than the one designed for matching which generates all sorts of extra partial nodes, as well as allowing callers to request other behavior changes using `vardef_assign` and the other flags. I think most callers don't actually need to visit generated partial nodes or do anything else strange, so I'm updating them to use the vanilla visitor instead. Where needed, I'll have callers inherit from the `matching_visitor` that currently backs `mk_visitor`.